In [1]:
import pandas as pd
from pythainlp.tokenize import word_tokenize
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.svm import LinearSVC
from sklearn.preprocessing import normalize

In [2]:
df = pd.read_csv('/Users/phrare/ds_class/wk2_nlp/truevoice-intent/mari_train.csv')
df.head()

texts  \
0   ผมไปจ่ายเงินที่ Counter Services เค้าเช็ต . บ...   
1               internet ยังความเร็วอยุ่เท่าไหร ครับ   
2   ตะกี้ไปชำระค่าบริการไปแล้ว แต่ยังใช้งานไม่ได้...   
3   พี่ค่ะยังใช้ internet ไม่ได้เลยค่ะ เป็นเครื่อ...   
4   ฮาโหล คะ พอดีว่าเมื่อวานเปิดซิมทรูมูฟ แต่มันโ...   

                                       texts_deepcut   action        object  \
0   ผม ไป จ่าย เงิน ที่ Counter Services เค้า เช็...  enquire       payment   
1          internet ยัง ความ เร็ว อยุ่ เท่า ไหร ครับ  enquire       package   
2   ตะกี้ ไป ชำระ ค่า บริการ ไป แล้ว แต่ ยัง ใช้ ...   report       suspend   
3   พี่ ค่ะ ยัง ใช้ internet ไม่ ได้ เลย ค่ะ เป็น...  enquire      internet   
4   ฮาโหล คะ พอดี ว่า เมื่อ วาน เปิด ซิม ทรูมูฟ แ...   report  phone_issues   

           destination  
0  billing and payment  
1           promotions  
2  billing and payment  
3             internet  
4  billing and payment

In [3]:
word_tokenize('สวัสดีครับ')

['สวัสดี', 'ครับ']

In [4]:
vectorizer = CountVectorizer(tokenizer=word_tokenize)

In [5]:
X_train = vectorizer.fit_transform(df['texts'])

In [6]:
words = vectorizer.get_feature_names()

In [7]:
len(words)
#words

3577

It is seen that there are total of 3577 words in our dictionary.

In [8]:
y_train = df['destination'].astype('category').cat.codes
y_train

0        0
1        5
2        0
3        2
4        0
5        4
6        4
7        0
8        4
9        2
10       0
11       4
12       2
13       4
14       4
15       0
16       4
17       0
18       0
19       0
20       0
21       2
22       0
23       2
24       0
25       4
26       0
27       4
28       0
29       0
        ..
12909    2
12910    2
12911    0
12912    0
12913    0
12914    5
12915    5
12916    5
12917    5
12918    0
12919    0
12920    5
12921    5
12922    5
12923    0
12924    0
12925    0
12926    2
12927    0
12928    5
12929    5
12930    0
12931    0
12932    0
12933    0
12934    0
12935    4
12936    0
12937    5
12938    5
Length: 12939, dtype: int8

In [9]:
clf = LinearSVC(random_state=0, tol=1e-5)

In [10]:
clf.fit(X_train, y_train)  

/Users/phrare/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


LinearSVC(C=1.0, class_weight=None, dual=True, fit_intercept=True,
          intercept_scaling=1, loss='squared_hinge', max_iter=1000,
          multi_class='ovr', penalty='l2', random_state=0, tol=1e-05,
          verbose=0)

In [11]:
df_test = pd.read_csv('/Users/phrare/ds_class/wk2_nlp/truevoice-intent/mari_test.csv')
df_test.head()

texts  \
0  อยากทราบว่าเหมาโทรทรูมูฟ 90 บาท 7 วัน รหัสอะไรค่ะ   
1  อยากทราบว่าโทรทัศน์ที่บ้านทำไมถึงดูไม่ได้ ดูได...   
2           อยากทราบว่าโปรเสริม 7 วัน ต้องกดอะไรครับ   
3    อยากทราบว่าโปรโมชัน 0863036995 เหลือเท่าไหร่ค่ะ   
4         อยากทราบว่าโปรโมชั่นจะหมดวันที่เท่าไหร่ค่ะ   

                                       texts_deepcut   action       object  \
0  อยาก ทราบ ว่า เหมา โทร ทรู มูฟ   90   บาท   7 ...  enquire    promotion   
1  อยาก ทราบ ว่า โทรทัศน์ ที่ บ้าน ทำไม ถึง ดู ไม...  enquire  nontruemove   
2  อยาก ทราบ ว่า โปรเสริม   7   วัน   ต้อง กด อะไ...      buy      package   
3  อยาก ทราบ ว่า โปรโมชัน   0863036995   เหลือ เท...  enquire      balance   
4  อยาก ทราบ ว่า โปรโมชั่น จะ หมด วัน ที่ เท่า ไห...  enquire      package   

           destination  
0           promotions  
1        other queries  
2           promotions  
3  billing and payment  
4           promotions

In [12]:
X_test = vectorizer.transform(df_test['texts'])

In [13]:
y_test = df_test['destination'].astype('category').cat.codes

In [14]:
clf.predict(X_test)

array([5, 4, 5, ..., 0, 0, 5], dtype=int8)

In [15]:
clf.score(X_test, y_test)

0.8498145859085291

In [23]:
X_train_norm = normalize(X_train, norm='l2')
X_test_norm = normalize(X_test, norm='l2')

In [24]:
clf.fit(X_train_norm,y_train)

LinearSVC(C=1.0, class_weight=None, dual=True, fit_intercept=True,
          intercept_scaling=1, loss='squared_hinge', max_iter=1000,
          multi_class='ovr', penalty='l2', random_state=0, tol=1e-05,
          verbose=0)

In [25]:
clf.predict(X_test_norm)

array([5, 4, 5, ..., 0, 0, 5], dtype=int8)

In [26]:
clf.score(X_test_norm, y_test)

0.8618665018541409

It is seen that with normalization, the score increases. :)